### PR TITLE
HDDS-2170. Add Object IDs and Update ID to Volume Object

### PR DESCRIPTION
**Pull request by anuengineer**
This patch proposes to add object ID and update ID when a volume is created.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 81 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 65 | Maven dependency ordering for branch |
| -1 | mvninstall | 42 | hadoop-hdds in trunk failed. |
| -1 | mvninstall | 28 | hadoop-ozone in trunk failed. |
| -1 | compile | 17 | hadoop-hdds in trunk failed. |
| -1 | compile | 13 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 59 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 927 | branch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 18 | hadoop-hdds in trunk failed. |
| -1 | javadoc | 16 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 1011 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 29 | hadoop-hdds in trunk failed. |
| -1 | findbugs | 16 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| -1 | mvninstall | 32 | hadoop-hdds in the patch failed. |
| -1 | mvninstall | 25 | hadoop-ozone in the patch failed. |
| -1 | compile | 20 | hadoop-hdds in the patch failed. |
| -1 | compile | 15 | hadoop-ozone in the patch failed. |
| -1 | cc | 20 | hadoop-hdds in the patch failed. |
| -1 | cc | 15 | hadoop-ozone in the patch failed. |
| -1 | javac | 20 | hadoop-hdds in the patch failed. |
| -1 | javac | 15 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 52 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 783 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 20 | hadoop-hdds in the patch failed. |
| -1 | javadoc | 20 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 28 | hadoop-hdds in the patch failed. |
| -1 | findbugs | 16 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 25 | hadoop-hdds in the patch failed. |
| -1 | unit | 20 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 2532 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1510 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 88621073dc38 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6cbe5d3 |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-mvninstall-hadoop-hdds.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-ozone.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-hdds.txt |
| cc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-hdds.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-findbugs-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/testReport/ |
| Max. process+thread count | 306 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/common hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1510/1/console |
| versions | git=2.7.4 maven=3.3.9 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeCreateRequest.java [hunk 1]
```diff
@@ -116,6 +116,11 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
     Collection<String> ozAdmins = ozoneManager.getOzoneAdmins();
     try {
       omVolumeArgs = OmVolumeArgs.getFromProtobuf(volumeInfo);
+      // when you create a volume, we set both Object ID and update ID to the
+      // same ratis transaction ID. The Object ID will never change, but update
+      // ID will be set to transactionID each time we update the object.
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeCreateRequest.java]: Do we need to add handling for updateID change in some volume metadata operations such as set/add/remove ACL, addMetadata, etc

Comment (anuengineer) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeCreateRequest.java]: yes, we need to. I wanted to make smaller patches. I will file and post the next patch as soon as this is committed.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmVolumeArgs.java [hunk 2]
```diff
@@ -188,6 +237,29 @@ public int hashCode() {
...
+    private long updateID;
+
+    /**
+     * Sets the Object ID for this Object.
+     * Object ID are unique and immutable identifier for each object in the
+     * System.
+     * @param objectID - long
+     */
+    public void setObjectID(long objectID) {
+      this.objectID = objectID;
```
Comment (bharatviswa504) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmVolumeArgs.java]: Just a question, why we need to have objectID and updateID?

Comment (anuengineer) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmVolumeArgs.java]: Object ID represents a unique Identity for an object.  Update ID represents a token that is useful to detect if an object has been changed after a read has been performed. For example, with update IDs I can cache an object and make changes, and when I apply I can detect if an object has been updated by any other thread.

**Comment by anuengineer** [comment]
Thanks for the reviews, @xiaoyuyao  and @bharatviswa504. Appreciate your time and careful thought.

**Pull request merged**